[apex] Add ApexCRUDViolation support for database class, inline no-arg object construction DML and inline list initialization DML

jonathanwiesel opened the pull request:
## Describe the PR

Until now `ApexCRUDViolation` didn't support checks when using the `Database` class DML methods, only the inline DML statements.

Also when a DML statement was issued with an inline instantiation of an SObject without any arguments it was not checked. Similarly with an inline instantiation of an SObject list with inline item adding.

## Related issues

<!-- PR relates to issues in the `pmd` repo: -->


- Fixes #3201

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: No java rules are changed!
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented on pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/security/xml/ApexCRUDViolation.xml:
```diff
@@ -134,6 +134,32 @@ public class Foo {
+    </test-code>
+
+    <test-code>
+        <description>No CRUD check for inline upsert with database class</description>
```
Could you extend this test case to include the sample code from #2628? Also please add the issue number in the description, this helps us later on to understand, why this test works as it works...

jonathanwiesel replied:
The example shown in #2628 is regarding querying records using a dynamic SOQL (string query) which is not being address here (although the comment regarding  DMLs does)

adangel replied:
Ah, I see. Yes, then, it's a different issue :) Thanks for clarifying.

adangel reviewed (commented):
Thanks for working on this!

The code looks good, I just have a few organizational requests:

* There are two commits in this PR, that look like they don't belong here: 5320d8e and 9a17cab - These are not needed, please remove them (e.g. rebase interactive).
* Can you verify, that this PR actually fixes the false-negative in #2628 ?
* You mentioned, you partially resolved #3201 - would you mind splitting this issue and separate the parts, that are solved now and put the missing part into a new issue? That way, we can close #3201 with this PR.

Could you please also have a look at the example code for the rule: https://pmd.github.io/latest/pmd_rules_apex_security.html#apexcrudviolation
I think, this example code is wrong: It checks for the Name column to be updateable, but actually the "Status__c" column is updated...
Do I understand this correct: If you use "WITH SECURITY_ENFORCED" in the query, you don't need to check, whether the table is accessible? (that's how the example code looks like).
The example code for the rule is here: 
https://github.com/pmd/pmd/blob/1b6fd08195e3fcb57c9e46873de27ccd1c8664ac/pmd-apex/src/main/resources/category/apex/security.xml#L54-L69

jonathanwiesel commented:
> There are two commits in this PR, that look like they don't belong here: 5320d8e and 9a17cab - These are not needed, please remove them (e.g. rebase interactive).

Fixed it

> Can you verify, that this PR actually fixes the false-negative in #2628 ?

Only DMLs yes, queries no.

> You mentioned, you partially resolved #3201 would you mind splitting this issue and separate the parts, that are solved now and put the missing part into a new issue? That way, we can close #3201 with this PR.

Absolutely, just split it to #3378 so #3201 can be fixed entirely.

> Could you please also have a look at the example code for the rule: https://pmd.github.io/latest/pmd_rules_apex_security.html#apexcrudviolation
I think, this example code is wrong: It checks for the Name column to be updateable, but actually the "Status__c" column is updated...

You're right, just updated it

> Do I understand this correct: If you use "WITH SECURITY_ENFORCED" in the query, you don't need to check, whether the table is accessible? (that's how the example code looks like).

Correct, the explicit check (like is done with the updating part would not be necessary since it will be automatically handed by the `WITH SECURITY_ENFORCED` keyword, a similar feature for DMLs is currently in pilot so in the short future would be easier to check for DMLs too.

adangel reviewed (approved):
Thanks!
